### ISSUE TITLE
Proposed feature to help "fit" displayables

### ISSUE THREAD

**Issue by mal** [renpy/renpy#2235]
Originates from discussion in https://github.com/renpy/renpy/pull/2230.

tl;dr: Current `maxsize` behaviour leaves the door open for confusion, and there are many related use-cases it would be nice to support. Below is a table of identified use-cases and a proposal based on the ideas shared in the original discussion.

<details>
<summary>Identified use-cases (click to expand)</summary>

![fit](https://user-images.githubusercontent.com/591257/84430861-3b2ea680-ac22-11ea-8de3-91c8b4a7db60.png)
</details>

---

The original discussion raises the prospect of a new `fit` property for transforms, that same concept is shown here but named `size_type` to more strongly link it with the `size` property but this is only a suggestion:

_Comments refer back to the identified use-cases._
```rpy
# size_type := (exact|fit|over|min|max) default: exact

# minsize - scales up until the smaller dimension (proportionally) matches or exceeds it's bound
size (300, 200)
size_type 'min'

# maxsize - scales down until the larger dimension (proportionally) matches or lies below it's bound
size (300, 200)
size_type 'max'

# fit x - scales proportionally until the width matches xsize
xsize 300
size_type 'fit'

# fit y - scales proportionally until the height matches ysize
ysize 200
size_type 'fit'

# fit inside - scales proportionally until the displayable fits inside the bounding box
size (300, 200)
size_type 'fit'

# fit over - scales proportionally until the displayable is able to cover the bounding box
size (300, 200)
size_type 'over'
```

The default would be current behaviour when setting `size`, with the introduction of `xsize` and `ysize` to facilitate only specifying a single dimension, and `size_type` to govern how the sizing is done. This would then allow the suggested deprecation of `minsize`/`maxsize` as both old and new cases should be covered. Additionally, if only one of `xsize`/`ysize` is set, the default `size_type` should probably become fit.

One limitation with this approach is that it wouldn't be possible to set separate min and max dimensions on the same transform, but I don't believe this is supported at the moment either. It could also be achieved relatively painlessly with one additional nested transform to allow for what is hopefully a relatively rare edge-case.

**Comment by Gouvernathor** [comment]
OK, so the basic behavior of `xsize` and `ysize` : I think the best would be to store `xsize` and `ysize` in `size`, overriding the preceding values if necessary. For example, something like :
`if size or xsize or ysize:`
    `size = (xsize or size[0], ysize or size[1])`
with `xsize`, `ysize` and `size` defaulting to None.
Therefore, writing `size=(234, 567), xsize=345, ysize=678` or `xsize=345, ysize=678` or `size=(789, 678), xsize=345` would result in the same as writing `size=(345, 678)`.

Then, the `size_type` property should not even be read if at least one of the two coordinates in `size` is missing, aka `if None in size`, and **in that case** `size_type` should be defaulting to "fit" (when the two are specified but `size_type` is missing, I agree it should be assumed as "exact", if only to match the current behavior for sake of compatibility).

I'll try to work around all that in code. Let me know if you disagree about software specifications.

**Comment by mal** [comment]
I've had some more (currently half-baked) thoughts around extra use-cases and ways to make it more intuitive, but they need a bit more mulling over. I should have time this weekend to try out the pull request, do some playing around of my own and iron out those ideas into something intelligible. Will report back then! 😁

**Comment by mal** [comment]
I put together a similar approach but following the pattern for other `xblah`/`yblah` properties (where `blah` is calculated rather than stored, and sought to support things like stretching (by using one of `xsize`/`ysize`  and `exact`), and determining the default `size_type` based on number of dimensions specified etc. `maxsize` is retained in it's current form for compatibility but is superseded if the new approach is also being used which seemed appropriate for a feature being nudged toward deprecation.

The `size_type` property should probably also be validated for expected values, however it wasn't immediately clear to be the correct approach to take, so I've left it open for now with the aim of testing it's functionality before solving that.

**Comment by mal** [comment]
Further investigation lead me to the `object-fit` property in the CSS3 spec, which has names for most of the things we want to support with this:

current | css3 | others
--------|------|---
`exact` | `fill` |
`min` | _no equivalent_ | `grow`, `expand`
`max` | `scale-down` | `shrink`
`fit` | `contain` |
`over` | `cover` |

Rather than trying to reinvent one I think it makes sense to mostly match this naming convention. I'm not quite convinced on `scale-up`/`scale-down`, but maybe that's more a period of adjustment thing, opinions on those would be great. They certainly an improvement over `min`/`max`. I also considered a few others which I noted. It would also probably make sense to go back to the originally proposed `fit` (or `size_fit`) for the property name governing this.

I'll see what feels most natural when putting together a test-case for this feature and report back, in the meantime any feedback would be welcome. 🙂

**Comment by mal** [comment]
PR (#2241) is now documented and tested. The test script and its output are included for review. I've gone with the CSS3-style names for now, but as per above we could tweak these if unhappy.

For the sake of completeness it's worth mentioning that during this process it did occur to me that reusing the `maximum` and `minimum` properties from the style system may be another way to approach this problem, however when thinking it through this quickly runs into issues when the starting aspect ratio is not known in advance, which would result in less flexibility and make it more error prone.

**Comment by renpytom** [comment]
#2241 has been merged, so this can be closed. Thanks everyone!
